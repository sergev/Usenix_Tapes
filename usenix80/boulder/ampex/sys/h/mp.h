/*
 * Definitions for Versatec Matrix printer/plotter driver
 *
 * see mpl.c & mp.c for driver proper.
 */
#define	NBUFS	5		/* maximum # of system buffers in use */
#define	MPOPRI	10		/* output priority (except during close) */
#define	MPLWAT	(NBUFS-1)*2	/* each buffer has at least one 'D' & 'l' command */
#define	MPHWAT	50
/*
 * Internal state.
 */
struct mp
  {	int m_flags;		/* see defines below */
	int *m_acs;		/* address of active csr */
	int m_cmd;		/* last command pulled from m_cmdq */
	int m_pri;		/* current sleep priority */
	int m_nbufs;		/* # of system buffers in use */
	int m_pagesize;		/* # of lines in page */
	int m_tofline;		/* # of lines from TOF to 1st print line */
	int m_ejline;		/* # of printing lines before auto-eject */
	int m_topeat;		/* # of input lines to eat at TOF */
	int m_vcc;		/* virtual column counter */
	int m_acc;		/* actual column counter */
	int m_alc;		/* actual line counter */
	int m_ilc;		/* input line counter */
	struct mpbuf *m_wbp;	/* top half's working buffer pointer */
	struct mpbuf *m_actf;	/* first buf in active buffer queue (pull) */
	struct mpbuf *m_actl;	/* last buf in active buffer queue (push) */
	struct clist m_tmpq;	/* temporary command queue */
	struct clist m_cmdq;	/* active command queue */
	struct clist m_canq;	/* canonical character processing queue */
  };
/*
 * mp.m_flags:
 */
#define	OPEN	01
#define	TEXT	02		/* plotter is in text mode (top half) */
#define	TMID	04		/* in middle of reading Text */
#define	RMID	010		/* in middle of reading Raster */
#define	IND	020		/* indent lines 1 tab (dev 1) */
#define PRINTER	040		/* one of the printer devices is open */
#define	ROLL	0100		/* roll paper is loaded */
#define	ERR	0200		/* error condition exists */
#define RETRY	0400		/* waiting for not READY to clear */
/*
 * Overlay template for buf struct
 */
struct mpbuf
  {	int m_flags;
	struct buf *m_forw;
	struct buf *m_back;
	struct mpbuf *m_next;	/* next in active queue */
	int m_used;		/* # of bytes used in buffer */
	dev_t m_dev;
	int m_wfree;		/* # of unallocated words in buffer */
	caddr_t m_addr;
#define m_extra m_un.m_ext
	union {
		int m_ext;
		daddr_t m_junk;
	}m_un;
	char m_xmem;
	char m_error;
	int m_resid;		/* # of raster bytes yet to be iomoved */
  };
/*
 * Format of raster and text data in buffer
 */
struct raster
  {	int r_len;		/* # of words in r_data[] */
	char r_data[];		/* raster dow row proper */
  };
/*
 * Versatec hardare defs
 */
#define MPRVECT	0174	/* raster interrupt vector */
#define	MPTVECT	0200	/* text interrupt vector */
#define	MPADDR	0177500

struct
  {	int mprbc;	/* raster byte count */
	int mpxmem;	/* dma memory extension */
	int mptbc;	/* text byte count */
	int mpma;	/* memory address */
	int mprcs;	/* raster control & status */
	int mprdr;	/* raster data register */
	int mptcs;	/* text control & status */
	int mptdr;	/* text data register */
  };
/*
 * csr bits:
 */
#define SPP	01	/* enable simultaneous print/plot */
#define RESET	02	/* remote reset */
#define CLRBUF	04	/* clear dot buffer */
#define EOT	010	/* remote end of transmission */
#define FF	020	/* remote form feed */
#define LINE	040	/* remote line terminate */
/* #define IENABLE	0100	/* interrupt enable for READY or ERROR */
#define READY	0200	/* matrix is ready for data byte */
#define DMABSY	020000	/* DMA data transfer in progress */
#define DTCIEN	040000	/* interrupt enable for DMA byte count == 0 */
#define ERROR	0100000	/* pwr off, no paper, off-line, non-exist mem */

/*
 * Misc.
 */
#define	FORM	014
#define ESC	033
#define	MAXCOL	132
#define	PAGESIZE 66		/* software conventional page length */
#define PAGELEN	68		/* actual # of lines in 8.5" page */
#define EJLINE	60		/* conventional # of printing lines (pr) */
#define EJAUTO	64		/* hardware # of printing lines */
#define TOPEAT	1		/* # of lines to discard at top in ESC mode */
#define TOFLINE	2		/* # conventional # of added lines at top */
#define TOFAUTO	2		/* # of lines from fanfold to auto eject */
#define	RSIZE	132		/* # of words in largest raster */
#define ROFFSET ((char *)&((struct raster *)0->r_data[0]))	/* offset of data
				   from start of raster structure */
