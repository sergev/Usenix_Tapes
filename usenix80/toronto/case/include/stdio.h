#define	BUFSIZ	512
#define	_NFILE	20
# ifndef FILE
extern	struct	_iobuf {
	char	*_ptr;
	int	_cnt;
	char	*_base;
	char	_flag;
	char	_file;
} _iob[_NFILE];
# endif

#define	_IOREAD	01
#define	_IOWRT	02
#define	_IONBF	04
#define	_IOMYBUF	010
#define	_IOEOF	020
#define	_IOERR	040
#define	_IOSTRG	0100
#define	NULL	0
#define	FILE	struct _iobuf
#define	EOF	(-1)

#define	stdin	(&_iob[0])
#define	stdout	(&_iob[1])
#define	stderr	(&_iob[2])
#define	getc(p)		(--(p)->_cnt>=0? *(p)->_ptr++&0377:_filbuf(p))
#define	getchar()	getc(stdin)
#define	putc(x,p)	(--(p)->_cnt>=0? (*(p)->_ptr++=(x)):_flsbuf((x),p))
#define	putchar(x)	putc(x,stdout)
#define	feof(p)		(((p)->_flag&_IOEOF)!=0)
#define	ferror(p)	(((p)->_flag&_IOERR)!=0)
#define	fileno(p)	p->_file

FILE	*fopen();
FILE	*freopen();

#define	_U	01
#define	_L	02
#define	_A	03
#define	_N	04
#define	_S	010

extern	char	_ctype[];

#define	isalpha(c)	(_ctype[c]&_A)
#define	isupper(c)	(_ctype[c]&_U)
#define	islower(c)	(_ctype[c]&_L)
#define	isdigit(c)	(_ctype[c]&_N)
#define	isspace(c)	(_ctype[c]&_S)
#define	toupper(c)	((c)-'a'+'A')
#define	tolower(c)	((c)-'A'+'a')
