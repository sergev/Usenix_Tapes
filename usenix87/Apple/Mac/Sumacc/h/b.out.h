/*	This file must be maintained in:
 *
 *		/trix/include/b.out.h
 *		/include/b.out.h
 *
 *	Layout of b.out file :
 *
 *	header of 8 longs	magic number 405, 407, 410, 411
 *				text size		)
 *				data size		) in bytes
 *				bss size		)
 *				symbol table size	)
 *				text relocation size	)
 *				data relocation size	)
 *				entry point
 *
 *
 *	header:			0
 *	text:			32
 *	data:			32+textsize
 *	symbol table:		32+textsize+datasize
 *	text relocation:	32+textsize+datasize+symsize
 *	data relocation:	32+textsize+datasize+symsize+rtextsize
 *
 */
/* various parameters */
#define SYMLENGTH	50		/* maximum length of a symbol */
#define SEGSIZE		32768		/* relocation boundry for 410 files */

/* types of files */
#define	ARCMAGIC 016230236041
#define OMAGIC	0405
#define	FMAGIC	0407
#define	NMAGIC	0410
#define	IMAGIC	0411
#define	RMAGIC	0700200000

/* symbol types */
#define	EXTERN	040
#define	UNDEF	00
#define	ABS	01
#define	TEXT	02
#define	DATA	03
#define	BSS	04
#define	COMM	05	/* internal use only */
#define REG	06

/* relocation regions */
#define	RTEXT	00
#define	RDATA	01
#define	RBSS	02
#define	REXT	03

/* relocation sizes */
#define RBYTE	00
#define RWORD	01
#define RLONG	02

/* macros which define various positions in file based on a bhdr, filhdr */
#define TEXTPOS		sizeof(filhdr)
#define DATAPOS 	TEXTPOS + filhdr.tsize
#define SYMPOS		DATAPOS + filhdr.dsize
#define RTEXTPOS	SYMPOS + filhdr.ssize
#define RDATAPOS	RTEXTPOS + filhdr.rtsize
#define ENDPOS		RDATAPOS + filhdr.rdsize
/* header of b.out files */
struct bhdr {
	long	fmagic;
	long	tsize;
	long	dsize;
	long	bsize;
	long	ssize;
	long	rtsize;
	long	rdsize;
	long	entry;
};

/* symbol management */
struct sym {
	char	stype;
	char	slength;
	long	svalue;
};

/* relocation commands */
struct reloc {
	unsigned rsegment:2;	/* RTEXT, RDATA, RBSS, or REXTERN */
	unsigned rsize:2;		/* RBYTE, RWORD, or RLONG */
	unsigned rdisp:1;		/* 1 => a displacement */
	short rsymbol;		/* id of the symbol of external relocations */
	long rpos;		/* position of relocation in segment */
};

