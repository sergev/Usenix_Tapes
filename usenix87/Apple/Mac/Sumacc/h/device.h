/*	device.h	1.0	07/21/84	*/

/*
 * Device driver definitions.
 */

/* 
 * Copyright (C) 1984, Stanford Univ. SUMEX project.  
 * May be used but not sold without permission.
 */

/*
 * history
 * 07/21/84	Croft	Created.
 * 07/27/84	Schuster Added accessory status codes.
 */

/*
 * Driver structure.  
 */
struct drvr {
	short	drvrFlags;	/* flags */
	short	drvrDelay;	/* ticks between actions */
	short	drvrEMask;	/* desk accessory event mask */
	short	drvrMenu;	/* menu ID */
	short	drvrOpen;	/* offset to open routine */
	short	drvrPrime;	/* ..		prime */
	short	drvrCtl;	/* ..		control */
	short	drvrStatus;	/* ..		status */
	short	drvrClose;	/* ..		close */
	char	drvrName[32];	/* driver name (pascal string) */
};

struct drvr drvr;		/* global instance of my drvr struct */

/* flags in drvrFlags */
#define	dReadEnable	0x100
#define	dWritEnable	0x200
#define	dCtlEnable	0x400
#define	dStatEnable	0x800
#define	dNeedGoodBye	0x1000
#define	dNeedTime	0x2000
#define	dNeedLock	0x4000
#define	dOpened		0x20
#define	dRAMBased	0x40
#define	drvrActive	0x80

/*
 * Commands in (low byte of) ioTrap.
 */
#define	aRdCmd	2
#define	aWrCmd	3
#define	aCtlCmd	4
#define	aStsCmd	5

/*
 * Special return codes from driver entry points.
 *
 * A zero or negative return from the driver routines causes an ioDone,
 * with the return status in D0.  The return values below (large 
 * positive numbers) cause the indicated action.
 */
#define	ioRTS 0x40000000
#define	ioKill	ioRTS
#define	ioDone	0

/*
 * Control and status codes.
 */
#define	killCode	1
#define	ejectCode	7
#define	drvStsCode	8

/*
 * IO System Errors (should be in osintf.h)
 */
#define	controlErr	-17
#define	statusErr	-18
#define	readErr		-19
#define	writErr		-20
#define	badUnitErr	-21
#define	unitEmptyErr	-22
#define	openErr		-23
#define	closErr		-24
#define	dRemovErr	-25
#define	dInstErr	-26
#define	abortErr	-27
#define	notOpenErr	-28


/*
 * Control and Status codes for desk accessories.
 */
#define accEvent 64
#define accRun 65
#define accCursor 66
#define accMenu 67
#define accUndo 68
#define accCut 70
#define accCopy 71
#define accPaste 72
#define accClear 73
