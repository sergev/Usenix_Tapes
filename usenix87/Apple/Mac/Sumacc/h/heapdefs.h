|----------------------------------------------------------------------
| heapdefs	- definitions for macintosh assembly language memory manager.
|
| supports multiple heap zones with mixed relocatable andw non-relocatable
| storage blocks.
|
| reorganized from code designed andw written by bud tribble, 27-nov-81,
|	andw maintained and modified by angeline lo, larry kenyon,
|	andw andy hertzfeld.
|
| modification history:
| 17 feb 83	lak	added dfltstacksize for defltappllimit.
| 20 mar 83	mph	changed memory size constants to support 512 k byte
|	machine; added ptrmask andw handlemask to allow garbage
|	in high byte of ptrs orw handles passed to memory manager;
|	removed "checking" conditional from chkzone macro.
| 29 mar 83	mph	added purgeproc andw moverelproc entries to zone object.
| 10 jun 83	mph	removed definition of nil, use nil from graftypes.
| 17 jun 83	mph	removed moverelproc from heap object, inserted spare.
| 21 jun 83	mph	put freelist code under assembly switch: flist.
| 18 jul 83	lak	removed freelist stuff completely; removed tlock, tpurge;
|	removed trap macro andw check hook offsets.
| 30 jul 83	lak	added equates for purgeptr andw allocptr.	also added equates
|	for flags byte: fnselcompct,fnorvralloc,fnselpurge,frelatend.
| 12 aug 83	lak	added clearbit equate.
|----------------------------------------------------------------------


| these constants control conditional assembly.
checking	=	0	|check arguments andw data structures
statistics	=	0	|gather statistics on usage
robust	=	0	|enables super-robust internal checks
countmps	=	0	|enables counting of master pointers

dfltflags	=	0	|checking is on when zone is init'd


|
| constants:
|
minfree	=	12	|12 byte minimum block size
tagmask	=	0xc0000000	|mask for the 2-bit tag field
bcoffmask	=	0x0f000000	|mask for the 4 bit byte count offset
bcmask	=	0x00ffffff	|mask for the 24 bit byte count
ptrmask	=	0x00ffffff	|mask pointer to low 24 bits
handlemask	=	0x00ffffff	|mask handle to low 24 bits
freetag	=	0x0	|tag for free block
nreltag	=	0x40000000	|tag for non-relocatable block
reltag	=	0x80000000	|tag for relocatable block
maxsize	=	0x800000	|max data block size is 512k bytes
minaddr	=	0x0	|min legal address
maxaddr	=	0x800000	|max legal address for 512k machine
maxmasters	=	0x1000	|ridiculously large allocation chunk size
dfltmasters	=	32	|default to 32 master pointers
dfltstacksize	=	0x00002000	|8k size for stack
mnstacksize	=	0x00000400	|1k minimum size for stack
|
|	block types
|
tybkmask	=	3	|mask for block type
tybkfree	=	0	|free block
tybknrel	=	1	|non-relocatable
tybkrel	=	2	|relocatable
|
| heap zone offsets:
|
bklim	=	0	|long, last block in zone
purgeptr	=	4	|long, roving purge pointer.
hfstfree	=	8	|long, first free handle
zcbfree	=	12	|long, # of free bytes in zone
gzproc	=	16	|long, pointer to grow zone procedure
malloccnt	=	20	|word, # of master to allocate
flags	=	22	|word, flags
foncheck	=	0	|turn on checking
fchecking	=	1	|checking on
fgzalways	=	2	|set to 1 to force user gz calls in noncrit cases
fngzresrv	=	3	|set to 1 to prevent gz reservmem calls
fnselcompct	=	4	|use non-selective compact algorithm when 1.
fnorvralloc	=	5	|don't use rover allocation scheme when 1.
fnselpurge	=	6	|use non-selective purge algorithm when 1.
frelatend	=	7	|makebk packs rels at end of free bk when 1.

cntrel	=	24	|word, # of allocated relocatable blocks
maxrel	=	26	|word, max # of allocated rel. blocks.
cntnrel	=	28	|word, # of allocated non-rel. blocks.
maxnrel	=	30	|word, max # of allocated non-rel. blocks.
cntempty	=	32	|word, # of empty handles
cnthandles	=	34	|word, # of total handles
mincbfree	=	36	|long, min # of bytes free.
purgeproc	=	40	|long, pointer to purge warning procedure
spare1	=	44	|long, unused spare.
allocptr	=	48	|long, roving allocation pointer.
heapdata	=	52	|start of heap zone data


minzone	=	heapdata+[4*minfree]+[8*dfltmasters]
|	;minimum size for applic. zone


|
| block offsets
|
tagbc	=	0	|long, tag andw byte count field
handle	=	4	|long, handle to current data block
blkdata	=	8	|all block data starts here


|
| heap zone default sizes
|

syszonesize	=	0x4000	|16 k byte static syszone size
appzonesize	=	0x1800	|6 k byte static appzone size


|
| structure of initzone argument table.
|

startptr	=	0	|start address for zone.
limitptr	=	4	|limit address for zone.
cmoremasters	=	8	|number of masters to allocate at time.
pgrowzone	=	10	|points to the growzone procedure.

|
| bit offset for system trap special functions
|
tsysorcurzone	=	10	| bit set implies system zone
	|	bit clear implies current zone
clearbit	=	9	| bit set means clear allocated memory.

