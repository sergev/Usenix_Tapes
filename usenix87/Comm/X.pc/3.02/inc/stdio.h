/**
*
* This header file defines the information used by the standard I/O
* package.
*
**/
#define _BUFSIZ 512		/* standard buffer size */
#define BUFSIZ 512		/* standard buffer size */
#define _NFILE 20		/* maximum number of files */

struct _iobuf
{
char *_ptr;			/* current buffer pointer */
int _rcnt;			/* current byte count for reading */
int _wcnt;			/* current byte count for writing */
char *_base;		 	/* base address of I/O buffer */
char _flag;			/* control flags */
char _file;			/* file number */
int _size;			/* size of buffer */
char _cbuff;			/* single char buffer */
char _pad;			/* (pad to even number of bytes) */
};

extern struct _iobuf _iob[_NFILE];

#define _IOREAD 1		/* read flag */
#define _IOWRT 2		/* write flag */
#define _IONBF 4		/* non-buffered flag */
#define _IOMYBUF 8		/* private buffer flag */
#define _IOEOF 16		/* end-of-file flag */
#define _IOERR 32		/* error flag */
#define _IOSTRG 64
#define _IORW 128		/* read-write (update) flag */

#if SPTR
#define NULL 0			/* null pointer value */
#else
#define NULL 0L
#endif
#define FILE struct _iobuf	/* shorthand */
#define EOF (-1)		/* end-of-file code */

#define stdin (&_iob[0])	/* standard input file pointer */
#define stdout (&_iob[1])	/* standard output file pointer */
#define stderr (&_iob[2])	/* standard error file pointer */

#define getc(p) (--(p)->_rcnt>=0? *(p)->_ptr++:_filbf(p))
#define getchar() getc(stdin)
#define putc(c,p) (--(p)->_wcnt>=0? ((int)(*(p)->_ptr++=(c))):_flsbf((c),p))
#define putchar(c) putc(c,stdout)
#define feof(p) (((p)->_flag&_IOEOF)!=0)
#define ferror(p) (((p)->_flag&_IOERR)!=0)
#define fileno(p) (p)->_file
#define rewind(fp) fseek(fp,0L,0)
#define fflush(fp) _flsbf(-1,fp)
#define clearerr(fp) clrerr(fp)

FILE *fopen();
FILE *freopen();
long ftell();
char *fgets();

#define abs(x) ((x)<0?-(x):(x))
#define max(a,b) ((a)>(b)?(a):(b))
#define min(a,b) ((a)<=(b)?(a):(b))

