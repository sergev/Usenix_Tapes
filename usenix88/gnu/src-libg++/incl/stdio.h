// This may look like C code, but it is really -*- C++ -*-
/* 
Copyright (C) 1988 Free Software Foundation
    written by Doug Lea (dl@rocky.oswego.edu)

This file is part of GNU CC.

GNU CC is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY.  No author or distributor
accepts responsibility to anyone for the consequences of using it
or for whether it serves any particular purpose or works at all,
unless he says so in writing.  Refer to the GNU CC General Public
License for full details.

Everyone is granted permission to copy, modify and redistribute
GNU CC, but only under the conditions described in the
GNU CC General Public License.   A copy of this license is
supposed to have been given to you along with GNU CC so you
can know your rights and responsibilities.  It should be in a
file named COPYING.  Among other things, the copyright notice
and this notice must be preserved on all copies.  
*/

/*
 *  Please check the following before installing this file:
 *
 *  If you are running AT&T System V or similar systems in which
 *  there is no _bufsiz field in _iobuf, then comment out the
 *  #define HAVE_BUFSIZ line below.
 *
 *  Check whether your libc.a sprintf function returns
 *  an int (as do most) versus a char* (BSD), and (un)comment
 *  the corresponding SPRINTF_RETURNS_INT line.
 *
 *  Check the value of BUFSIZ against the one in your /usr/include/stdio.h.
 *
 *  Carefully check the fields and order of _iobuf declaration against
 *  the one in your /usr/include/stdio.h. Xenix-based systems
 *  may need some re-ordering of _iobuf. fields.
 *
 *  Note that some _IOXXX #defines may not be present in your 
 *  /usr/include/stdio.h. This is ok, so long as the ones that
 *  are present in both are set to the same values.
 *
 *  Some of the prototypes refer to functions that may not be
 *  present in your libc.a. This is ok so long as you do not
 *  actually call such functions.
 *
 */

#ifndef FILE

/* check and possibly comment out the following */
#define HAVE_BUFSIZ 
#define SPRINTF_RETURNS_INT

/* check and possibly redefine the following */
#define BUFSIZ  1024            

extern  struct  _iobuf {
    int      _cnt;
    char*    _ptr;
    char*    _base;
#   ifdef HAVE_BUFSIZ
    int     _bufsiz;
    short   _flag;
#   else
    char    _flag;
#   endif
    char    _file;
} _iob[];

#define FILE     struct _iobuf

#define _IOFBF    00000
#define _IOREAD   00001
#define _IOWRT    00002
#define _IONBF    00004
#define _IOMYBUF  00010
#define _IOEOF    00020
#define _IOERR    00040
#define _IOSTRG   00100
#define _IOLBF    00200
#define _IORW     00400
#define _IOAPPEND 01000

#define EOF       (-1)

#ifndef NULL
#define NULL      0
#endif

#define stdin     (&_iob[0])
#define stdout    (&_iob[1])
#define stderr    (&_iob[2])

#define getc(p) (--(p)->_cnt>=0?(int)(*(unsigned char*)(p)->_ptr++):_filbuf(p))
#define putc(x,p) (--(p)->_cnt>=0? ((int)(*(p)->_ptr++=(unsigned)(x))):_flsbuf((unsigned)(x),p))

#define clearerr(p) ((p)->_flag &= ~(_IOERR|_IOEOF))
#define getchar()   getc(stdin)
#define putchar(x)  putc(x,stdout)
#define feof(p)     (((p)->_flag&_IOEOF)!=0)
#define ferror(p)   (((p)->_flag&_IOERR)!=0)
#define fileno(p)   ((p)->_file)

extern int    _doprnt(const char*, void*, FILE*);
extern int    _doscan(FILE*, const char*, void*);
extern int    _filbuf(FILE*);
extern int    _flsbuf(unsigned, FILE*);
extern int    fclose(FILE*);
extern FILE*  fdopen(int, const char*);
extern int    fflush(FILE*);
extern int    fgetc(FILE*);
extern char*  fgets(char*, int, FILE *);
extern FILE*  fopen(const char*, const char*);
extern int    fprintf(FILE*, const char* ...);
extern int    fputc(int, FILE*);
extern int    fputs(const char*, FILE*);
extern int    fread(void*, int, int, FILE*);
extern FILE*  freopen(const char*, const char*, FILE*);
extern int    fscanf(FILE*, const char* ...);
extern int    fseek(FILE*, long, int);
extern long   ftell(FILE *);
extern int    fwrite(const void*, int, int, FILE*);
extern char*  gets(char*);
extern int    getw(FILE*);
extern int    pclose(FILE*);
extern FILE*  popen(const char*, const char*);
extern int    printf(const char* ...);
extern void   puts(const char*);
extern int    putw(int, FILE*);
extern int    scanf(const char* ...);
extern void   setbuf(FILE*, char*);
extern void   setbuffer(FILE*, char*, int);
extern void   setlinebuf(FILE*);
extern void   setvbuf(FILE*, char*, int, int);
extern int    sscanf(char*, const char* ...);
extern FILE*  tmpfile();
extern int    ungetc(int, FILE*);
extern int    vfprintf(FILE*, const char*, void* ap);
extern int    vprintf(const char*, void* ap);
extern int    vsprintf(char*, const char*, void* ap);

#ifdef SPRINTF_RETURNS_INT
extern int    sprintf(char*, const char* ...);
#else
extern char*  sprintf(char*, const char* ...);
#endif

#endif // FILE
