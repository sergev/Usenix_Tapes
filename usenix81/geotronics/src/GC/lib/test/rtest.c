/*
	rtest -- test random-number library routines

	last edit: 03-Jan-1981	D A Gwyn
*/

#include	<stdio.h>

extern double	RED(), RND(), RandFlt();
extern int	RPD(), RandInt();

#define	TRIALS	400

#define	MAX	40
static int	count[MAX];		// "bins" to tally "hits"

main()	{
	register int	i , k ;

	/* test random exponential deviates */
	printf( "\n\nRandom exponential deviates:\n\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		count[i] = 0 ;		// clear counters
	for ( k = 0 ; k < TRIALS ; ++ k )
		{
		i = 10.0 * RED();	// truncated
		if ( i < 0 )
			i = 0 ;
		else if ( i >= MAX )
			i = MAX - 1 ;
		++ count[i];		// tally hit
		}
	Histo();			// display histogram

	/* test random normal deviates */
	printf( "\f\n\nRandom normal deviates:\n\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		count[i] = 0 ;		// clear counters
	for ( k = 0 ; k < TRIALS ; ++ k )
		{
		i = 20.0 + 5.0 * RND();	// truncated
		if ( i < 0 )
			i = 0 ;
		else if ( i >= MAX )
			i = MAX - 1 ;
		++ count[i];		// tally hit
		}
	Histo();			// display histogram

	/* test random Poisson deviates */
	printf( "\f\n\nRandom Poisson deviates:\n\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		count[i] = 0 ;		// clear counters
	for ( k = 0 ; k < TRIALS ; ++ k )
		{
		i = RPD( 10.0 );	// truncated
		if ( i < 0 )
			i = 0 ;
		else if ( i >= MAX )
			i = MAX - 1 ;
		++ count[i];		// tally hit
		}
	Histo();			// display histogram

	/* test random uniform floating-point deviates */
	printf( "\f\n\nRandom uniform floating-point deviates:\n\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		count[i] = 0 ;		// clear counters
	for ( k = 0 ; k < TRIALS ; ++ k )
		{
		i = RandFlt( 10.0 , 30.0 );	// truncated
		if ( i < 0 )
			i = 0 ;
		else if ( i >= MAX )
			i = MAX - 1 ;
		++ count[i];		// tally hit
		}
	Histo();			// display histogram

	/* test random uniform integer deviates */
	printf( "\f\n\nRandom uniform integer deviates:\n\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		count[i] = 0 ;		// clear counters
	for ( k = 0 ; k < TRIALS ; ++ k )
		{
		i = RandInt( 10 , 30 );
		if ( i < 0 )
			i = 0 ;
		else if ( i >= MAX )
			i = MAX - 1 ;
		++ count[i];		// tally hit
		}
	Histo();			// display histogram
	}
/*
	Histo - plot histogram of hit counts
*/

Histo()	{
	register int	i , j ;

	printf( "# ....5...10...15...20...25...30...35...40\n" );
	for ( i = 0 ; i < MAX ; ++ i )
		{
		printf( "%2d" , i );
		for ( j = 0 ; j < count[i] ; ++ j )
			putchar( `*' );
		putchar( `\n' );
		}
	}
