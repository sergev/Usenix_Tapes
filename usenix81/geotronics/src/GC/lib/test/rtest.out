

Random exponential deviates:

# ....5...10...15...20...25...30...35...40
 0********************************
 1**********************************
 2*******************************
 3**************************
 4************************
 5*************************
 6**********************
 7****************
 8*****************
 9******************
10******************
11**********
12*********
13*******
14**************
15*********
16**********
17********
18*********
19******
20*******
21*******
22*****
23****
24**
25**
26**
27
28*
29***
30
31***
32***
33***
34*
35**
36**
37*
38***
39****


Random normal deviates:

# ....5...10...15...20...25...30...35...40
 0
 1
 2
 3
 4
 5
 6
 7**
 8**
 9****
10*****
11*****
12**********
13************
14**************
15*********************
16********************
17****************************
18*****************************
19************************
20*************************************
21*********************************
22******************************
23*******************************
24***********************
25********************
26**********
27**************
28**********
29************
30*
31
32*
33
34*
35*
36
37
38
39


Random Poisson deviates:

# ....5...10...15...20...25...30...35...40
 0
 1
 2*
 3*****
 4*************
 5**********
 6**********************
 7**************************
 8**********************************************************
 9**********************************************
10*************************************************
11*************************************************
12********************************
13********************************
14**************************
15*************
16**********
17*****
18**
19*
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39


Random uniform floating-point deviates:

# ....5...10...15...20...25...30...35...40
 0
 1
 2
 3
 4
 5
 6
 7
 8
 9
10**********************
11******************
12***********************
13************************
14**********************
15**********************
16*****************
17*******************************
18***************
19*************
20**********************
21***************
22********************
23******************
24*************************
25***************
26**********************
27*****************
28**********************
29*****************
30
31
32
33
34
35
36
37
38
39


Random uniform integer deviates:

# ....5...10...15...20...25...30...35...40
 0
 1
 2
 3
 4
 5
 6
 7
 8
 9
10************************
11***********************
12******************
13***************
14******************
15***************
16************************
17*********************
18*******************
19*******************
20***************
21*******************
22*****************
23*****************
24************
25**********************
26*******************
27****************
28**********************************
29***************
30******************
31
32
33
34
35
36
37
38
39
